gc/gc-iceberg/src/main/java/org/projectnessie/gc/iceberg/ExpireSnapshotsProcedure.java

```diff
@@ -0,0 +1,435 @@
+          .forEach(
+              entry -> {
+                // Expiry Algorithm:
+                // A. For each content id, fetch ANY one of the content.
+                // B. Use the newly created expiry branch. On this branch,
+                // C. Do a single PUT operation with a fetched content from step-A and
+                //    a dummy ContentKey.
+                // D. Collect the list of all the expired snapshot for this content id.
+                // E. Fetch this table using dummy ContentKey on expiry branch. For this table,
+                // F. call expire_snapshot action from Iceberg by passing the list of expired
+                //    snapshots.
+                // G. Once all the content ids are handled, drop this branch.
+                //    (Else it may impact the next GC operations as some contents are still referred
+                // in
+                //    this branch)
+                String contentId = entry.getKey();
+                addOneContentToExpireProcedureBranch(
+                    expiryBranchName, identifiedResultsRepo, api, finalRunId, contentId);
+                Table table =
+                    nessieCatalog.loadTable(
+                        TableIdentifier.parse(EXPIRE_NAMESPACE + "." + contentId));
+
+                // TODO: Remove after https://github.com/apache/iceberg/pull/4509
+                updateMetadataLocation((BaseTable) table);
+
+                Set<Snapshot> snapshotSet =
+                    new TreeSet<>((o1, o2) -> (int) (o1.timestampMillis() - o2.timestampMillis()));
+                Set<String> processedMetadataFiles = new HashSet<>();
+                List<String> metadataLocations =
+                    new ArrayList<>(metadataLocationsPerContentId.get(contentId));
+
+                metadataLocations.sort(Collections.reverseOrder());
+
+                metadataLocations.forEach(
+                    metadataLocation -> {
+                      if (!processedMetadataFiles.contains(metadataLocation)) {
+                        TableMetadata metadata =
+                            TableMetadataParser.read(table.io(), metadataLocation);
+                        snapshotSet.addAll(metadata.snapshots());
+                        List<String> files =
+                            metadata.previousFiles().stream()
+                                .map(TableMetadata.MetadataLogEntry::file)
+                                .collect(Collectors.toList());
+                        processedMetadataFiles.addAll(files);
+                      }
+                    });
+
+                // create a dummy table metadata which has all the snapshots for a content from all
+                // the references.
+                TableMetadata base = ((BaseTable) table).operations().current();
+                base = TableMetadata.buildFrom(base).build();
+                try {
+                  FieldUtils.writeField(base, "previousFiles", Collections.emptyList(), true);
```

Comment:
Strong -1 on relying on Iceberg implementation details + using error-prone hacks!

Reply:
Dummy metadata has to have same UUID as the content's any of the metadata. 
There is no other interface from Iceberg to do that other than using the base object. 

Also due to lot of internal validations in table metadata, I had to do this.
I don't find other way.

Reply:
You can create TableMetadata from scratch, no?

Reply:
Nope. That interface uses random UUID inside and it won't match with the table's UUID. So, some validations will fail.

Reply:
Also `new TableMeatdata()` is not public constructor. It is package private

Reply:
Used a `GcMetadataUtil` with same package name as iceberg to prepare a metadata without history to avoid using reflections.